content/news/2025-08-04-gsoc-2025-report-armaan-chowfin.md

```diff
@@ -1,17 +1,232 @@
-title: GSoC 2025 Work Product - Resampling Options for Mixxx
+title: GSOC 2025 Work Product - Resampling Options for Mixxx
 authors: Armaan Chowfin
 tags: gsoc, gsoc-2025
 comments: yes
 date: 2025-08-04 17:42:53
 
-This GSOC project is derived from [Mixxx issue#9328](https://github.com/mixxxdj/mixxx/issues/9328). The goal is to determine whether alternative interpolation algorithms result in a *noticeable reduction in scratching artifacts or a latency improvement over linear interpolation*, **providing quantitative supporting evidence** where possible.
+Presenting the latest additions to Mixxx:  
+**[issue#9328](https://github.com/mixxxdj/mixxx/issues/9328): Additional interpolation options for scratching**  
+**[issue#10611](https://github.com/mixxxdj/mixxx/issues/10611): Custom output samplerates for recording and broadcast**.
+
+The scratching engine now supports a sinc-based resampler with three quality settings, and the original linear resampler is now twice as fast. Additionally, outbound audio is resampled before encoding for recording or streaming. This can improve compatibility with limited or fixed-rate sound cards.
 
 #### Introduction
-On a turntable, scratching is performed by moving the stylus by hand - causing it to follow grooves in the vinyl that correspond to the analog audio waveform. Mixxx offers the ability to emulate vinyl scratching on digital records, by spinning the jog wheels of a MIDI controller to emulate the motion of the turntable stylus during scratching. The action of scratching causes a sudden acceleration or deceleration in tempo of the loaded record(s).
+Mixxx uses the `SoundTouch` and `RubberBand` time-stretching libraries for resampling during a keyLock operation. However, these libraries are unsuitable for scratching due to the fast changing tempo and pitch. Currently a faster, handcrafted linear interpolation algorithm is used - but there have been reports of suboptimal audio quality.  
+
+Digital Signal Processing (DSP) theory tells us that linear interpolation is not ideal, and that a sinc-based resampler will always return interpolated values identical to the original analog signal, under certain theoretical constraints. However, practical implementations of sinc resampling are computationally heavy and generally unsuitable for low-latency realtime software such as Mixxx. Therefore,
+
+* One objective of this GSoC project was to explore the feasibility of using sinc interpolation for scratching. To this end, the `libsamplerate` and `libzita` resample latencies were evaluated.
+* Another focus was to investigate and improve the performance of the current linear resampler. Here, we observed that the `libsamplerate` linear interpolator outperformed our own, *reducing per-buffer resample latency from 20µs to 10µs.*
```

Comment:
Please use space between numbers and units See: https://physics.nist.gov/cuu/Units/checklist.html